test/e2e/framework/endpoints/ports.go

```diff
@@ -0,0 +1,135 @@
+limitations under the License.
+*/
+
+/*
+This soak tests places a specified number of pods on each node and then
+repeatedly sends queries to a service running on these pods via
+a serivce
+*/
+
+package endpoints
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
+)
+
+// PortsByPodName is a map that maps pod name to container ports.
+type PortsByPodName map[string][]int
+
+// PortsByPodUID is a map that maps pod UID to container ports.
+type PortsByPodUID map[types.UID][]int
+
+// GetContainerPortsByPodUID returns a PortsByPodUID map on the given endpoints.
+func GetContainerPortsByPodUID(ep *v1.Endpoints) PortsByPodUID {
+	m := PortsByPodUID{}
+	for _, ss := range ep.Subsets {
+		for _, port := range ss.Ports {
+			for _, addr := range ss.Addresses {
+				containerPort := port.Port
+				if _, ok := m[addr.TargetRef.UID]; !ok {
+					m[addr.TargetRef.UID] = make([]int, 0)
+				}
+				m[addr.TargetRef.UID] = append(m[addr.TargetRef.UID], int(containerPort))
+			}
+		}
+	}
+	return m
+}
+
+func translatePodNameToUIDOrFail(c clientset.Interface, ns string, expectedEndpoints PortsByPodName) PortsByPodUID {
+	portsByUID := make(PortsByPodUID)
+
+	for name, portList := range expectedEndpoints {
+		pod, err := c.CoreV1().Pods(ns).Get(name, metav1.GetOptions{})
+		if err != nil {
+			framework.Failf("failed to get pod %s, that's pretty weird. validation failed: %s", name, err)
+		}
+		portsByUID[pod.ObjectMeta.UID] = portList
+	}
+	// Logf("successfully translated pod names to UIDs: %v -> %v on namespace %s", expectedEndpoints, portsByUID, ns)
+	return portsByUID
+}
+
+func validatePortsOrFail(ep PortsByPodUID, expectedEndpoints PortsByPodUID) {
+	if len(ep) != len(expectedEndpoints) {
+		// should not happen because we check this condition before
+		framework.Failf("invalid number of endpoints got %v, expected %v", ep, expectedEndpoints)
+	}
+	for podUID := range expectedEndpoints {
+		if _, ok := ep[podUID]; !ok {
+			framework.Failf("endpoint %v not found", podUID)
+		}
+		if len(ep[podUID]) != len(expectedEndpoints[podUID]) {
+			framework.Failf("invalid list of ports for uid %v. Got %v, expected %v", podUID, ep[podUID], expectedEndpoints[podUID])
+		}
+		sort.Ints(ep[podUID])
+		sort.Ints(expectedEndpoints[podUID])
+		for index := range ep[podUID] {
+			if ep[podUID][index] != expectedEndpoints[podUID][index] {
+				framework.Failf("invalid list of ports for uid %v. Got %v, expected %v", podUID, ep[podUID], expectedEndpoints[podUID])
+			}
+		}
+	}
+}
+
+// ValidateEndpointsOrFail validates that the given service exists and is served by the given expectedEndpoints.
+func ValidateEndpointsOrFail(c clientset.Interface, namespace, serviceName string, expectedEndpoints PortsByPodName) {
```

Comment:
Can we rename this to `ValidateEndpointsPortsOrFail`?

Reply:
> Keep 2 commits for now, if good to merge, I'll squash them to 1 commit

it is better to squash if just two commits and very small.